flash_io.c

```diff
@@ -0,0 +1,622 @@
+  {
+    if(flash_ptr_1.ptr[f1_index].data.tag                 == TAG &&
+       flash_ptr_1.ptr[f1_index].data.targets.as_unsigned == target.as_unsigned)
+    {
+      valid = VALID_1;
+    }
+  }
+
+  if(f2_index != -1)
+  {
+    if(flash_ptr_2.ptr[f2_index].data.tag                 == TAG &&
+       flash_ptr_2.ptr[f2_index].data.targets.as_unsigned == target.as_unsigned)
+    {
+      if(valid == VALID_1)
+        valid = BOTH_VALID;
+      else
+        valid = VALID_2;
+    }
+  }
+
+  return valid;
+}
+
+static void
+flash_update_lookup(flash_valid_data_t* lookup, furnace_context_t* ctx)
+{
+  lookup->log_bits    = ctx->log_bits;
+  lookup->pwm_level   = ctx->pwm_level;
+  lookup->ceiling_pwm = ctx->ceiling_pwm;
+
+#if CONFIG_WATER
+  lookup->pwm_water = ctx->pwm_water;
+#endif
+
+#if CONFIG_MAGNETRON
+  lookup->magnetron_pulse_count = ctx->pulse_count;
+  lookup->magnetron_deadline    = ctx->magnetron_deadline;
+#endif
+
+#if CONFIG_AUTO == CONFIG_AUTO_PILOT || CONFIG_AUTO == CONFIG_AUTO_MAPPER
+  lookup->pilot_is_enabled = ctx->pilot.is_enabled;
+  lookup->pilot_des_temp   = ctx->pilot.des_temp;
+#endif
+
+#if CONFIG_AUTO == CONFIG_AUTO_MAPPER
+  lookup->mapper_is_enabled   = ctx->mapper.is_enabled;
+  lookup->mapper_max_pwm_temp = ctx->mapper.max_pwm_temp;
+#endif
+
+}
+
+static void
+flash_write_(flash_ptr_t* flash_ptr)
+{
+  const uint32_t ints = save_and_disable_interrupts();
+  flash_range_program((uint32_t)&flash_ptr->ptr[flash_ptr->current_index] - XIP_BASE,
+                       (uint8_t*)&flash_current, FLASH_PAGE_SIZE);
+  restore_interrupts(ints);
+}
+
+static void
+flash_write()
+{
+  if(flash_ptr_1.current_index == flash_ptr_2.current_index)
+  {
+    flash_ptr_1.current_index++;
+    flash_ptr_1.current_index %= NUM_PAGES_PER_SECTOR;
+    flash_write_(&flash_ptr_1);
+
+    if(flash_ptr_2.current_index == FLASH_MAX_PAGE_INDEX)
+      flash_clear_sector(flash_ptr_2.ptr);
+  }
+  else
+  {
+    flash_ptr_2.current_index++;
+    flash_ptr_2.current_index %= NUM_PAGES_PER_SECTOR;
+    flash_write_(&flash_ptr_2);
+
+    if(flash_ptr_1.current_index == FLASH_MAX_PAGE_INDEX)
+      flash_clear_sector(flash_ptr_1.ptr);
+  }
+}
+
+void
+init_flash(furnace_context_t* ctx)
+{
+  flash_find_last_blocks();
+
+  switch(flash_validate_blocks())
+  {
+    case VALID_1:
+    {
+      flash_read(&flash_ptr_1, ctx);
+    } break;
+
+    case VALID_2:
+    {
+      flash_read(&flash_ptr_2, ctx);
+    } break;
+
+    case BOTH_VALID:
+    {
+      if(flash_ptr_1.current_index == flash_ptr_2.current_index)
+        flash_read(&flash_ptr_2, ctx);
+      else
+        flash_read(&flash_ptr_1, ctx);
+    } break;
+  }
+
+  flash_update_lookup(&flash_last_written, ctx);
+  flash_update_lookup(&flash_current.data, ctx);
+
+  ctx->flash_deadline = make_timeout_time_ms(FLASH_WRITE_MS);
+}
+
+void
+do_flash_work(furnace_context_t* ctx)
+{
+  const absolute_time_t now = get_absolute_time();
+  if(now < ctx->flash_deadline)
+    return;
```

Comment:
Perhaps we should check some global flag that the config changed and should be flashed here.
The flag would be set if we update any field that is saved in the flash.

Reply:
We would have to remember to set this flag every time we change something in the config. Wouldn't it be better to simply perform a `memcmp` every `FLASH_WRITE_MS`? In #18, I proposed this feature with a flag, and you suggested automatically checking for changes, so I'm proposing this design. If you have a better solution in mind, please explain.